### PR TITLE
Handle incorrect address errors in terminal location API

### LINKED ISSUE #6122: Validate city and state values for correctness in terminal locations
### Describe the bug

We use [terminal locations](https://stripe.com/docs/terminal/fleet/locations) to group card readers for in-person payments. The creation of a terminal location in our systems is managed internally (no explicit API calls required) when the [`get_store_location`](https://github.com/Automattic/woocommerce-payments/blob/26c07c409f3753614442549931dfd7219d4be282/includes/admin/class-wc-rest-payments-terminal-locations-controller.php#L159) endpoint is called.

When creating a location today, we check to make sure that the city and state fields are not empty. Soon, Stripe will start validating city in all locations and state in countries (where it is required) for correctness (see [Stripe docs](https://stripe.com/docs/terminal/fleet/locations#create-location)).

We need to handle the appropriate error response from Stripe so that our existing API clients (eg. Woo mobile app) do not break.

### Additional context

* Internal discussion - pbUcTB-1Z2-p2#comment-6404
* [Stripe docs detailing validations for address fields](https://stripe.com/docs/terminal/fleet/locations#create-location)

### PR DESCRIPTION
**Pull request by anu-rock**
Fixes #6122

#### Changes proposed in this Pull Request

Re-route the new upstream error code `invalid_request_error` for incorrect city and state fields as our known error code `store_address_is_incomplete` to ensure existing mobile app versions are aware of the new validations.

#### Testing instructions

In order to trigger the correctness validation error from Stripe, we need to send an invalid city, state, or country. At the time of creating this PR, I found only an invalid state or country triggered an error. Specifying an incorrect state or country is impossible in WC settings, so we'll have to make a hacky test.

##### Regular test

* Go to WC settings at **WP-Admin** → **WooCommerce** → **Settings**, enter an invalid city name and click "Save changes".
* Make a GET request to the `/wp-json/wc/v3/payments/terminal/locations/store` endpoint. Refer to the [docs](https://github.com/woocommerce/woocommerce/wiki/Getting-started-with-the-REST-API) to set up authorization correctly.
* Before: `invalid_request_error` was returned in response. Now: `store_address_is_incomplete` is returned in response.

##### Hacky test

If your site doesn't already have an attached terminal location, the above call will create one using the store address in WC settings. Otherwise, your existing (and probably cached) location will be returned.

If the regular test ends up returning an existing location rather than creating a new one or if city validation doesn't work for you, try this temporary method:

* Comment out [these lines of code](https://github.com/Automattic/woocommerce-payments/blob/be3c4cf62ee8ebaa91a208f5007b47322fcc74e1/includes/admin/class-wc-rest-payments-terminal-locations-controller.php#L149-L157) to force creating a new location.
* Hard-code an incorrect state or country [here](https://github.com/Automattic/woocommerce-payments/blob/be3c4cf62ee8ebaa91a208f5007b47322fcc74e1/includes/admin/class-wc-rest-payments-terminal-locations-controller.php#L115-L122). For instance:
```php
'country'     => 'UnknownCountry',
```
* Make a GET request to the `/wp-json/wc/v3/payments/terminal/locations/store` endpoint. Refer to the [docs](https://github.com/woocommerce/woocommerce/wiki/Getting-started-with-the-REST-API) to set up authorization correctly.
* Before: `invalid_request_error` was returned in response. Now: `store_address_is_incomplete` is returned in response.

##### Success screenshot

image

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5989` or branch name `update/city-state-validation-terminal-location` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=update/city-state-validation-terminal-location) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 255de0133c695f3ef5e1208d5ab47506f18de70d
- Build time: 2023-04-27 08:29:59 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php [hunk 1]
```diff
@@ -162,7 +162,22 @@ public function get_store_location( $request ) {
 
 			return rest_ensure_response( $this->extract_location_fields( $location ) );
 		} catch ( API_Exception $e ) {
-			return rest_ensure_response( new WP_Error( $e->get_error_code(), $e->getMessage() ) );
+			$error = new WP_Error( $e->get_error_code(), $e->getMessage() );
+			if ( 'city_invalid' === $e->get_error_code() || 'state_invalid' === $e->get_error_code() ) {
```
Comment (kidinov) [inline includes/admin/class-wc-rest-payments-terminal-locations-controller.php]: The approach looks good to me! It looks like the thought that we have to match on `"param": "address[state/city/...]"`?

Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php]: It turns out that `param` being a custom field is not included in `API_Exception`. The best we can do is handle the error code "invalid_request_error" returned by Stripe for invalid state or country.

Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php]: I'll make this PR ready for review tomorrow with proper description and testing instructions. Other than that, I am not expecting more code changes.

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php [hunk 2]
```diff
@@ -162,7 +162,27 @@ public function get_store_location( $request ) {
...
 			return rest_ensure_response( $this->extract_location_fields( $location ) );
 		} catch ( API_Exception $e ) {
-			return rest_ensure_response( new WP_Error( $e->get_error_code(), $e->getMessage() ) );
+			$error = new WP_Error( $e->get_error_code(), $e->getMessage() );
+			// Stripe will return a 400 for incorrect city, state, or country.
+			// Ideally, we should return a more appropriate error code like 'store_address_is_incorrect',
+			// but that will break older mobile app clients.
+			// Until we have a more granular versioning support for WCPay REST endpoints,
+			// this is the best we can do.
+			if ( 'invalid_request_error' === $e->get_error_code() ) {
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-terminal-locations-controller.php:169]: Can we add additional verifications by checking if the current city is empty (see 'get_store_location' for accessing the address). With that, we can differentiate the target case from anything else.

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php [hunk 3]
```diff
@@ -162,7 +162,27 @@ public function get_store_location( $request ) {
 
 			return rest_ensure_response( $this->extract_location_fields( $location ) );
 		} catch ( API_Exception $e ) {
-			return rest_ensure_response( new WP_Error( $e->get_error_code(), $e->getMessage() ) );
+			$error = new WP_Error( $e->get_error_code(), $e->getMessage() );
+			// Stripe will return a 400 for incorrect city, state, or country.
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-terminal-locations-controller.php]: Please compact the comment into 2-3 lines if possible.

**Review by kalessil — CHANGES_REQUESTED** [review]

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php (hunk 2 above)
Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php:169]: You mean something like this?

```php
if ( isset( $location_address['city'] ) && 'invalid_request_error' === $e->get_error_code() )
```

Do you think the check on this line may not always be reliable?

https://github.com/Automattic/woocommerce-payments/blob/88e01940e3e0a5a4c23c35645d3bf269b85f6770/includes/admin/class-wc-rest-payments-terminal-locations-controller.php#L127

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php (hunk 3 above)
Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php]: Sure, done in 0c80f6d.

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php (hunk 2 above)
Comment (kalessil) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php:169]: Rather `if ( 'invalid_request_error' === $e->get_error_code() && ! isset( $location_address['city'] ) )`.

It should be reliable, but we check different error code here (invalid_request_error). Adding also city check will guarantee it's the issue with the address and not anything else.

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php (hunk 3 above)
Comment (kalessil) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php]: Thanks!

File: includes/admin/class-wc-rest-payments-terminal-locations-controller.php (hunk 2 above)
Comment (anu-rock) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php:169]: Ok, I think I got the essence of your suggestion. But `invalid_request_error` is returned by Stripe not only when there are missing address fields but also when a field like city, state or country has a value but one that's incorrect/invalid. IMO, that makes it hard to put additional checks.

For a bit more context, this is what Stripe returns for incorrect address fields:

```json
{
  "error": {
    "message": "Invalid US state or province",
    "param": "address[state]",
    "request_log_url": "https://dashboard.stripe.com/test/logs/req_ABC?t=123",
    "type": "invalid_request_error"
  }
}
```

And this is what it returns for missing address fields:

```json
{
  "error": {
    "code": "parameter_missing",
    "doc_url": "https://stripe.com/docs/error-codes/parameter-missing",
    "message": "Missing required address field for a Location in US: address[city].",
    "request_log_url": "https://dashboard.stripe.com/test/logs/req_ABC?t=123",
    "type": "invalid_request_error"
  }
}
```

Ideally, the `param` and `code` properties in the above response should be used along with the invalid_request_error check for better reliability. However, `API_Exception` object doesn't forward these fields to us. All it gives us is:

```json
{
    "code": "invalid_request_error",
    "message": "Invalid US state or province",
    "data": null
}
```

So, unless we tweak the `API_Exception` or `Exception` class, these are all the properties we can depend on :(

Or at least that's my understanding.

Comment (kalessil) [reply includes/admin/class-wc-rest-payments-terminal-locations-controller.php:169]: Gotha! Then I'm withdrawing my suggestion for adding more checks here =)

**Review by kalessil — APPROVED** [review]
LGTM!

**Review by elazzabi — APPROVED** [review]
LGTM 👍

**Pull request merged**